### PR TITLE
Perform I-Cache Maintenance on RISC-V.

### PR DESCRIPTION
**Pull request by yuyang-ok**
This is trying to fix https://github.com/bytecodealliance/wasmtime/issues/5033.
Thanks @afonso360, I think he is done all of the investigation,I didn't do a lot of work.
But I am still have question about the flag parameter.
look like `SYS_RISCV_FLUSH_ICACHE_ALL` and `SYS_RISCV_FLUSH_ICACHE_LOCAL` have the same value.

### CODE REVIEW

File: crates/jit-icache-coherence/src/libc.rs [hunk 1]
```diff
@@ -60,6 +60,38 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+                },
+                // Current the parameter is not used,But defined. https://github.com/torvalds/linux/blob/4d1044fcb996e8de9b9ab392f4a767890e45202d/arch/riscv/kernel/sys_riscv.c#L64-L66
+                0, // start
+                0, // end
+                {
+                    #[allow(non_snake_case)]
+                    let SYS_RISCV_FLUSH_ICACHE_LOCAL = 1;
+                    #[allow(non_snake_case)]
+                    let SYS_RISCV_FLUSH_ICACHE_ALL = SYS_RISCV_FLUSH_ICACHE_LOCAL;
+                    SYS_RISCV_FLUSH_ICACHE_ALL
```
Comment (afonso360) [inline crates/jit-icache-coherence/src/libc.rs:87]: I'm also a bit confused as to why the kernel decide to alias both of these settings, but it seems as it was just to represent the state of things at [the time](https://github.com/torvalds/linux/commit/921ebd8f2c081b3cf6c3b29ef4103eef3ff26054). I think we should be OK just referencing `SYS_RISCV_FLUSH_ICACHE_ALL` which is the operation that we want but adding a comment on the current situation.

File: crates/jit-icache-coherence/src/libc.rs [hunk 2]
```diff
@@ -60,6 +60,38 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
         // In any other case we got an actual error, so lets propagate that up
         e => e?,
     }
+    #[cfg(all(target_arch = "riscv64", target_os = "linux"))]
+    if num_cpus::get() == 1 {
```
Comment (bjorn3) [inline crates/jit-icache-coherence/src/libc.rs:64]: `num_cpus::get()` does multiple syscalls and as such may be more expensive than the `SYS_RISCV_FLUSH_ICACHE_ALL` syscall it is meant to avoid.

Comment (yuyang-ok) [reply crates/jit-icache-coherence/src/libc.rs:64]: Ok,My main consideration is that detected at runtime make it is more easy to use. use this library need no configuration.Maybe I will add a feature `one-core`. Because of multi-core is the most case.

File: crates/jit-icache-coherence/src/libc.rs [hunk 3]
```diff
@@ -60,6 +60,38 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+    } else {
+        match unsafe {
+            libc::syscall(
+                {
+                    // libc don't define the special syscall,So define it by My own.
+                    // https://github.com/torvalds/linux/search?q=__NR_arch_specific_syscall
+                    #[allow(non_snake_case)]
+                    let __NR_arch_specific_syscall = 244;
+                    __NR_arch_specific_syscall + 15
+                },
```
[the remark this answers is not in the document]
Comment (yuyang-ok) [reply crates/jit-icache-coherence/src/libc.rs:78]: yeah.

File: crates/jit-icache-coherence/src/libc.rs [hunk 4]
```diff
@@ -60,6 +60,38 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+                // Current the parameter is not used,But defined. https://github.com/torvalds/linux/blob/4d1044fcb996e8de9b9ab392f4a767890e45202d/arch/riscv/kernel/sys_riscv.c#L64-L66
+                0, // start
+                0, // end
```
[the remark this answers is not in the document]
Comment (yuyang-ok) [reply crates/jit-icache-coherence/src/libc.rs:81]: ok.

**Comment by yuyang-ok** [comment]
@afonso360 @bjorn3  I think we are ready.

File: crates/jit-icache-coherence/src/libc.rs [hunk 5]
```diff
@@ -77,12 +120,14 @@ fn membarrier(barrier: libc::c_int) -> Result<()> {
...
-pub(crate) fn clear_cache(_ptr: *const c_void, _len: usize) -> Result<()> {
+pub(crate) fn clear_cache(ptr: *const c_void, len: usize) -> Result<()> {
     // TODO: On AArch64 we currently rely on the `mprotect` call that switches the memory from W+R to R+X
     // to do this for us, however that is an implementation detail and should not be relied upon
     // We should call some implementation of `clear_cache` here
     //
     // See: https://github.com/bytecodealliance/wasmtime/issues/3310
 
+    #[cfg(all(target_arch = "riscv64", target_os = "linux"))]
+    riscv_flush_icache(ptr as u64, (ptr as u64) + (len as u64));
```
Comment (afonso360) [inline crates/jit-icache-coherence/src/libc.rs:131]:
```suggestion
    riscv_flush_icache(ptr as u64, (ptr as u64) + (len as u64))?;
```

File: crates/jit-icache-coherence/src/libc.rs [hunk 6]
```diff
@@ -60,10 +60,53 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
         // In any other case we got an actual error, so lets propagate that up
         e => e?,
     }
+    #[cfg(all(target_arch = "riscv64", target_os = "linux"))]
+    return riscv_flush_icache(0, 0);
```
Comment (afonso360) [inline crates/jit-icache-coherence/src/libc.rs:64]: I don't think we need to do this here. With the way the API for this crate is set up, we should always have a `clear_cache` operation preceding this. So we are effectively doing the cache flushing twice!

File: crates/jit-icache-coherence/src/libc.rs (hunk 5 above)
Comment (yuyang-ok) [reply crates/jit-icache-coherence/src/libc.rs:131]: ok.

File: crates/jit-icache-coherence/src/libc.rs (hunk 6 above)
Comment (yuyang-ok) [reply crates/jit-icache-coherence/src/libc.rs:64]: ok,thanks.

**Comment by yuyang-ok** [comment]
@bjorn3 @afonso360  I think we are ready.

**Comment by afonso360** [comment]
This looks correct to me.

File: crates/jit-icache-coherence/Cargo.toml [hunk 7]
```diff
@@ -20,4 +20,9 @@ features = [
 ]
 
 [target.'cfg(any(target_os = "linux", target_os = "macos"))'.dependencies.libc]
-version = "0.2.42"
\ No newline at end of file
+version = "0.2.42"
+
+[features]
+# Most modern cpu are smp,when This is enable some certain optimization can perform.
+# like riscv can use `fence.i` instead of linux syscall `sys_riscv_flush_icache`.
```
Comment (cfallin) [inline crates/jit-icache-coherence/Cargo.toml:27]: Writing/grammar fix: let's reword this as

```
# Most modern CPUs are SMP (multicore). However, when only one core is present,
# some aspects of coherence are much cheaper. For example, RISC-V can use
# one instruction `fence.i` rather than a syscall that invokes all other cores.
# This feature enables such optimizations, but the resulting program will *only*
# be safe to run on one-core systems.
```

File: crates/jit-icache-coherence/Cargo.toml [hunk 8]
```diff
@@ -20,4 +20,9 @@ features = [
...
+one-core = []
```
Comment (cfallin) [inline crates/jit-icache-coherence/Cargo.toml:28]: Missing newline after this line

File: crates/jit-icache-coherence/src/libc.rs [hunk 9]
```diff
@@ -64,6 +64,47 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+#[cfg(all(target_arch = "riscv64", target_os = "linux"))]
+fn riscv_flush_icache(start: u64, end: u64) -> Result<()> {
+    cfg_if::cfg_if! {
+        if #[cfg(feature = "one-core")] {
+            use std::arch::asm;
+            unsafe {
+                asm!("fence.i");
+            };
+            return Ok(())
+        }else {
```
Comment (cfallin) [inline crates/jit-icache-coherence/src/libc.rs:76]: Missing space before `else`

File: crates/jit-icache-coherence/src/libc.rs [hunk 10]
```diff
@@ -64,6 +64,47 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+            match unsafe {
+                libc::syscall(
+                    {
+                        // libc don't define the special syscall,So define it by My own.
```
Comment (cfallin) [inline crates/jit-icache-coherence/src/libc.rs:80]:
```suggestion
                        // The syscall isn't defined in `libc`, so we definfe the syscall number here.
```

File: crates/jit-icache-coherence/src/libc.rs [hunk 11]
```diff
@@ -64,6 +64,47 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+                        // https://github.com/torvalds/linux/search?q=__NR_arch_specific_syscall
+                        #[allow(non_upper_case_globals)]
+                        const  __NR_arch_specific_syscall :i64 = 244;
+                        // https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/tools/arch/riscv/include/uapi/asm/unistd.h#L40
+                        #[allow(non_upper_case_globals)]
+                        const sys_riscv_flush_icache :i64 =  __NR_arch_specific_syscall + 15;
+                        sys_riscv_flush_icache
+                    },
+                    // Current the parameter is not used,But defined. https://github.com/torvalds/linux/blob/4d1044fcb996e8de9b9ab392f4a767890e45202d/arch/riscv/kernel/sys_riscv.c#L64-L66
```
Comment (cfallin) [inline crates/jit-icache-coherence/src/libc.rs:89]:
```suggestion
                    // Currently these parameters are not used, but they are still defined.
```

File: crates/jit-icache-coherence/src/libc.rs [hunk 12]
```diff
@@ -64,6 +64,47 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
 
+#[cfg(all(target_arch = "riscv64", target_os = "linux"))]
+fn riscv_flush_icache(start: u64, end: u64) -> Result<()> {
+    cfg_if::cfg_if! {
+        if #[cfg(feature = "one-core")] {
+            use std::arch::asm;
+            unsafe {
+                asm!("fence.i");
+            };
+            return Ok(())
```
Comment (cfallin) [inline crates/jit-icache-coherence/src/libc.rs:75]: `return` isn't necessary here, I think: the whole `if` is an expression that will return the last value in the block.

File: crates/jit-icache-coherence/src/libc.rs [hunk 13]
```diff
@@ -64,6 +64,47 @@ pub(crate) fn pipeline_flush_mt() -> Result<()> {
...
+                        #[allow(non_snake_case)]
+                        const SYS_RISCV_FLUSH_ICACHE_LOCAL :i64 = 1;
+                        #[allow(non_snake_case)]
+                        const SYS_RISCV_FLUSH_ICACHE_ALL :i64 = SYS_RISCV_FLUSH_ICACHE_LOCAL;
+                        SYS_RISCV_FLUSH_ICACHE_ALL
+                    }, // flags
+                )
+            } {
+                0 => {return Ok(()) }
+                _ => return Err(Error::last_os_error()),
```
Comment (cfallin) [inline crates/jit-icache-coherence/src/libc.rs:102]: LIkewise for these two lines: `return` isn't needed in either case.

**Review by cfallin — COMMENTED** [review]
Thanks! I have some code-cleanup comments, but the logic looks fine to me too. OK to merge once these are all addressed.

File: crates/jit-icache-coherence/Cargo.toml (hunk 7 above)
Comment (alexcrichton) [reply crates/jit-icache-coherence/Cargo.toml:27]: I may have missed some discussion on this, but how come this is a crate feature? It seems like this would be better detected by the crate dynamically since many folks compiling code aren't certain what the target system looks like.

Comment (afonso360) [reply crates/jit-icache-coherence/Cargo.toml:27]: There a comment about detecting the number of cpu's at runtime: https://github.com/bytecodealliance/wasmtime/pull/5078#discussion_r1000502204

I'd prefer if we could call the [vDSO implementation](https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/arch/riscv/kernel/vdso/flush_icache.S#L10), since that would also do the exact same thing for non SMP kernels, and avoid the syscalls.

Comment (alexcrichton) [reply crates/jit-icache-coherence/Cargo.toml:27]: Could the result of `num_cpus::get` be stored in a `static`? That would incur syscalls on the first hit but none subsequently.

Comment (yuyang-ok) [reply crates/jit-icache-coherence/Cargo.toml:27]: @alexcrichton That is a good idea save `num_cpus::get`  in `static`.

File: crates/jit-icache-coherence/src/libc.rs (hunk 9 above)
Comment (yuyang-ok) [reply crates/jit-icache-coherence/src/libc.rs:76]: ok.

**Comment by yuyang-ok** [comment]
I think I am in favour of store `num_cpu` in static. @cfallin @bjorn3 @afonso360

**Comment by yuyang-ok** [comment]
I am having this error.
~~~
yuyang@yuyang-Inspiron-7590:~/projects/wasmtime/crates/jit-icache-coherence$ cargo t    --target   riscv64gc-unknown-linux-gnu 
   Compiling wasmtime-jit-icache-coherence v2.0.0 (/home/yuyang/projects/wasmtime/crates/jit-icache-coherence)
error[E0015]: cannot call non-const fn `num_cpus::get` in statics
  --> crates/jit-icache-coherence/src/libc.rs:12:35
   |
12 | static NUM_CPUS: usize = unsafe { num_cpus::get() };
   |                                   ^^^^^^^^^^^^^^^
   |
   = note: calls in statics are limited to constant functions, tuple structs and tuple variants

For more information about this error, try `rustc --explain E0015`.
error: could not compile `wasmtime-jit-icache-coherence` due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `wasmtime-jit-icache-coherence` due to previous error
~~~
In rust I have to use `lazy_static` to assign  this value??

File: crates/jit-icache-coherence/Cargo.toml (hunk 7 above)
Comment (bjorn3) [reply crates/jit-icache-coherence/Cargo.toml]: I don't think `num_cpus::get()` is safe. Linux supports hotplugging entire cpus and enabling/disabling cpu cores at runtime. In addition cgroups can be used to limit and unlimit cpu usage. `num_cpus` considers a cgroup limit to 1 or less cpu worth of time as there being only a single core even if it can migrate to a different cpu.

Comment (yuyang-ok) [reply crates/jit-icache-coherence/Cargo.toml]: @bjorn3  I have rollback to features implementation since not safe,safety come first.

**Comment by yuyang-ok** [comment]
@bjorn3 @afonso360 @alexcrichton @cfallin I think we are ready.

**Comment by yuyang-ok** [comment]
@cfallin  Are we going to merge this or not?

**Review by cfallin — APPROVED** [review]
@yuyang-ok , sorry about this falling through the cracks! It looks good to me. If you can rebase, I'm happy to merge.

**Comment by cfallin** [comment]
@yuyang-ok unfortunately this seems to be failing a number of tests now -- bad merge?
